### ISSUE TITLE
Regex interpolation performance regression

### ISSUE THREAD

**Issue by AlexDaniel** [rakudo/rakudo#2537]
Code:
```perl6
my $r = q|\d+|; "asdf123fdsa321" ~~ /<$r>/ for ^300_000; say now - INIT now
```

Bisected: (2018-07-19) https://github.com/rakudo/rakudo/commit/51e41da8d0d9b946f9c20426e29e5bc909e49dc5

Within that bump there are these moar commits: https://github.com/MoarVM/MoarVM/compare/2018.06-238-gfae91a6eb...2018.06-262-g0b08d3c18

The timings I get are somewhat noisy, but it seems to be noticeably slower:
```
<committable6> AlexDaniel, ¦51e41da8d0^: «8.4672288␤» ¦51e41da: «9.6204246␤»
<committable6> AlexDaniel, ¦51e41da8d0^: «8.8887845␤» ¦51e41da: «9.6929083␤»
<committable6> AlexDaniel, ¦51e41da8d0^: «8.80442657␤» ¦51e41da: «9.0541159␤»
<committable6> AlexDaniel, ¦51e41da8d0^: «8.23307304␤» ¦51e41da: «9.2722514␤»
```

**Comment by lizmat** [comment]
I think this spans the period where @jnthn did the scalar refactor, as described in https://6guts.wordpress.com/2018/07/27/redesigning-rakudos-scalar/

FWIW, I wonder if this slowdown could be caused by slower dynamic variable lookup.

**Comment by jnthn** [comment]
I've both recovered the performance lost in the regression and achieved a further speedup for this benchmark, helping regex applications even without interpolation a bit along the way.

The regressed time on my machine was 5.31s. The regression was noisy but around the 13% mark. After various optimizations, it now runs in 3.87s for me, which is about 73% of the time before I started fixing it. So I figure I've gained a good bit more than was lost.
